p2p/simulations/adapters/exec.go

```diff
@@ -307,21 +367,54 @@ func execP2PNode() {
 		}
 	}
 
-	// start the devp2p stack
-	stack, err := startP2PNode(&conf.Stack, services)
+	// initialize the devp2p stack
+	stack, err := node.New(&conf.Stack)
 	if err != nil {
-		log.Crit("error starting p2p node", "err", err)
+		log.Crit("error creating node stack", "err", err)
 	}
 
-	// use stdin / stdout for RPC to avoid the parent needing to access
-	// either the local filesystem or TCP stack (useful when running in
-	// Docker)
-	handler, err := stack.RPCHandler()
-	if err != nil {
-		log.Crit("error getting RPC server", "err", err)
+	// register the services, collecting them into a map so we can wrap
+	// them in a snapshot service
+	services := make(map[string]node.Service, len(serviceNames))
+	for _, name := range serviceNames {
+		serviceFunc, exists := serviceFuncs[name]
+		if !exists {
+			log.Crit("unknown node service", "name", name)
+		}
+		constructor := func(nodeCtx *node.ServiceContext) (node.Service, error) {
+			ctx := &ServiceContext{
+				RPCDialer:   &wsRPCDialer{addrs: conf.PeerAddrs},
+				NodeContext: nodeCtx,
+				Config:      conf.Node,
+			}
+			if conf.Snapshots != nil {
+				ctx.Snapshot = conf.Snapshots[name]
+			}
+			service, err := serviceFunc(ctx)
+			if err != nil {
+				return nil, err
+			}
+			services[name] = service
+			return service, nil
+		}
+		if err := stack.Register(constructor); err != nil {
+			log.Crit("error starting service", "name", name, "err", err)
+		}
 	}
-	go handler.ServeCodec(rpc.NewJSONCodec(&stdioConn{os.Stdin, os.Stdout}), rpc.OptionMethodInvocation|rpc.OptionSubscriptions)
 
+	// register the snapshot service
+	if err := stack.Register(func(ctx *node.ServiceContext) (node.Service, error) {
+		return &snapshotService{services}, nil
+	}); err != nil {
+		log.Crit("error starting snapshot service", "err", err)
+	}
+
+	// start the stack
+	if err := stack.Start(); err != nil {
+		log.Crit("error stating node stack", "err", err)
+	}
+
+	// stop the stack if we get a SIGTERM signal
 	go func() {
 		sigc := make(chan os.Signal, 1)
 		signal.Notify(sigc, syscall.SIGTERM)
```

Comment:
We only need to catch `SIGTERM` because that is what the ExecAdapter sends the process to stop it (see [here](https://github.com/ethersphere/go-ethereum/blob/network-testing-framework/p2p/simulations/adapters/exec.go#L199)).

If the process does get an unexpected `SIGINT` it will just do the default thing and just die, and you can't catch a `SIGKILL` :wink:.